File: Java/rockPaperScissors/RockPaperScissors.java
import java.io.*;
import java.util.*;

public class RockPaperScissors {

	public static void main(String[] args) throws IOException {
		IO io = new IO(System.in);
		int nPlayers, nGames, p1, p2;
		String p1move, p2move;
		boolean first_run = true;

		while(true){
		    nPlayers = io.nextInt();
		    if (nPlayers == 0) {
		        break;
		    }
		    if (first_run) {
		        first_run = false;
		    } else {
		    	System.out.printf("\n");
		    }
			nGames = io.nextInt();
		    int[] resultsW = new int[nPlayers];
		    int[] resultsL = new int[nPlayers];
		    for (int i = 0; i < ( ((nGames*nPlayers)*(nPlayers - 1)) / 2 ) ; i++) {
		    	p1 = io.nextInt();
		    	p1move = io.next();
		    	p2 = io.nextInt();
		    	p2move = io.next();
		        if (p1move.equals(p2move)){
		            continue;
		        }
		        if ((p1move.equals("scissors") && p2move.equals("paper")) || (p1move.equals("paper") && p2move.equals("rock")) || (p1move.equals("rock") && p2move.equals("scissors"))) {
		            resultsW[p1 - 1]++;
		            resultsL[p2 - 1]++;
		        } else {
		            resultsW[p2 - 1]++;
		            resultsL[p1 - 1]++;
		        }
		    } 
		    for (int i = 0; i < nPlayers; i++) {
		        double w_plus_l = resultsL[i] + resultsW[i];
		        if  (w_plus_l == 0){
		            System.out.println("-");
		        } else {
		            System.out.printf("%.3f\n", (resultsW[i] / w_plus_l));
		        }
		    }
		}
		
		io.close();
	}

	static class IO extends PrintWriter {
		static BufferedReader r;
		static StringTokenizer t;

		public IO(InputStream i) {
			super(new BufferedOutputStream(System.out));
			r = new BufferedReader(new InputStreamReader(i));
			t = new StringTokenizer("");
		}

		public String next() throws IOException {
			while (!t.hasMoreTokens()) {
				t = new StringTokenizer(r.readLine());
			}
			return t.nextToken();
		}

		public int nextInt() throws IOException{
			return Integer.parseInt(next());
		}

		public long nextLong() throws IOException {
			return Long.parseLong(next());
		}

		public double nextDouble() throws IOException {
			return Double.parseDouble(next());
		}
	}
}

File: Python/rockpaperscissors/rockpaperscissors.py
# WORKS BUT ISN'T FAST ENOUGH
first_run = True
while(True):
    inp = input().split()
    if len(inp) == 1:
        break
    if first_run:
        first_run = False
    else:
        print()
    nPlayers, nGames = [int(x) for x in inp]
    resultsW = [0] * nPlayers
    resultsL = [0] * nPlayers
    for i in range( int( ((nGames*nPlayers)*(nPlayers - 1)) / 2 ) ):
        p1, p1move, p2, p2move = [int(x) if x.isdigit() else x for x in input().split()]
        if p1move == p2move:
            continue
        if (p1move == "scissors" and p2move == "paper") or (p1move == "paper" and p2move == "rock") or (p1move == "rock" and p2move == "scissors"):
            resultsW[p1-1] += 1
            resultsL[p2-1] += 1
        else:
            resultsW[p2-1] += 1
            resultsL[p1-1] += 1

    for i in range(nPlayers):
        w_plus_l = resultsL[i] + resultsW[i]
        if  w_plus_l == 0:
            print("-")
        else:
            print("%.3f" % (resultsL[i] / w_plus_l))
    print("\n\n\n\n\n\n\n")
    print(resultsW)